sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExecHelper.scala

```diff
@@ -0,0 +1,303 @@
+        expr.map(JoinStateKeyWatermarkPredicate)
+
+      } else if (isWatermarkDefinedOnInput) { // case 2 explained in the class docs
+        val stateValueWatermark = getStateValueWatermark(
+          attributesToFindStateWatemarkFor = oneSideInputAttributes,
+          attributesWithEventWatermark = otherSideInputAttributes,
+          condition,
+          eventTimeWatermark)
+        val inputAttributeWithWatermark = oneSideInputAttributes.find(_.metadata.contains(delayKey))
+        val expr = watermarkExpression(inputAttributeWithWatermark, stateValueWatermark)
+        expr.map(JoinStateValueWatermarkPredicate)
+
+      } else {
+        None
+
+      }
+    }
+
+    val leftStateWatermarkPredicate =
+      getOneSideStateWatermarkPredicate(leftAttributes, leftKeys, rightAttributes)
+    val rightStateWatermarkPredicate =
+      getOneSideStateWatermarkPredicate(rightAttributes, rightKeys, leftAttributes)
+    JoinStateWatermarkPredicates(leftStateWatermarkPredicate, rightStateWatermarkPredicate)
+  }
+
+  /**
+   * Get state value watermark (see [[StreamingSymmetricHashJoinExec]] for context about it)
+   * given the join condition and the event time watermark. This is how it works.
+   * - The condition is split into conjunctive predicates, and we find the predicates of the
+   *   form `leftTime + c1 < rightTime + c2`   (or <=, >, >=).
+   * - We canoncalize the predicate and solve it with the event time watermark value to find the
+   *  value of the state watermark.
+   *
+   * @param attributesToFindStateWatemarkFor attributes of the side whose state watermark
+   *                                         is to be calculated
+   * @param attributesWithEventWatermark  attributes of the other side which has a watermark column
+   * @param joinCondition                 join condition
+   * @param eventWatermark                watermark defined on the input event data
+   * @return state value watermark in milliseconds
+   */
+  def getStateValueWatermark(
+      attributesToFindStateWatemarkFor: Seq[Attribute],
+      attributesWithEventWatermark: Seq[Attribute],
+      joinCondition: Option[Expression],
+      eventWatermark: Option[Long]): Option[Long] = {
+    if (joinCondition.isEmpty || eventWatermark.isEmpty) return None
+
+    def getStateWatermarkSafely(l: Expression, r: Expression): Option[Long] = {
+      try {
+        getStateWatemarkFromLessThenPredicate(
+          l, r, attributesToFindStateWatemarkFor, attributesWithEventWatermark, eventWatermark)
+      } catch {
+        case NonFatal(e) =>
+          logWarning(s"Error trying to extract state constraint from condition $joinCondition", e)
+          None
+      }
+    }
+
+    val allStateWatermarks = splitConjunctivePredicates(joinCondition.get).flatMap { predicate =>
+      val stateWatermark = predicate match {
+        case LessThan(l, r) => getStateWatermarkSafely(l, r)
+        case LessThanOrEqual(l, r) => getStateWatermarkSafely(l, r).map(_ - 1)
```

Comment:
why `-1`?

Reply:
the generated the state watermark cleanup expression is inclusive of the state watermark. That is if state watermark is W, all state where timestamp <= W will be cleaned up. 
Now when the canonicalized join condition solves to leftTime >= W, then I dont want to clean up <= W. Rather I choose to cleanup <= W-1.

Reply:
added this as docs.